js/ui/paygUnlockDialog.js

```diff
@@ -358,6 +362,56 @@ var PaygUnlockDialog = new Lang.Class({
         return buttonsBox;
     },
 
+    _onCodeSuccesfullyAdded(mainBox) {
+        mainBox.remove_all_children();
+
+        // successMessage will handle the formatting of the string
+        let timeString = Payg.successMessage(Main.paygManager._lastTimeAdded);
+        let successLabel = new St.Label({ style_class: 'unlock-dialog-payg-success',
+                                    text: timeString,
+                                    x_align: Clutter.ActorAlign.CENTER });
+        mainBox.add_child(successLabel);
+
+        let successButtonBox = new St.BoxLayout({ style_class: 'unlock-dialog-payg-button-box',
+                                        vertical: false,
+                                        x_expand: true,
+                                        x_align: Clutter.ActorAlign.CENTER,
+                                        y_expand: true,
+                                        y_align: Clutter.ActorAlign.END });
+
+        this._successButton = new St.Button({ style_class: 'modal-dialog-button button',
+                                         button_mask: St.ButtonMask.THREE,
+                                         reactive: true,
+                                         can_focus: true,
+                                         label: _("Success!"),
+                                         x_align: St.Align.MIDDLE,
+                                         y_align: St.Align.END });
+
+        this._successButton.add_style_pseudo_class('default');
+        successButtonBox.add_child(this._successButton);
+        mainBox.add_child(successButtonBox);
+        this._successButton.grab_key_focus();
+
+        this._timeoutId = Mainloop.timeout_add_seconds(3, () => {
+            Main.paygManager.emit('success-message-shown');
+
+            return GLib.SOURCE_REMOVE;
+        });
+
+        this._successButton.connect('button-press-event', () => {
+            this._onButtonConnect();
+        });
+
+        this._successButton.connect('key-press-event', () => {
+            this._onButtonConnect();
+        });
+    },
+
+    _onButtonConnect() {
+            Mainloop.source_remove(this._timeoutId);
+            Main.paygManager.emit('success-message-shown');
```

Comment:
I still don't like this design where this bit of UI tells some other object to emit a signal. If no unlock dialog is being displayed when expiry-time-changed is emitted, what's supposed to happen?

Reply:
Hmm I did this because the screenshield depended already on the emission of the signal emitted by the paygManager, and since both objects have a reference to the manager, i though it would be ok to use it to make it communicate across the code. 

The idea that the unlock dialog is not being displayed, it won't be able to capture that signal because it was killed by the screenshield, right?

Reply:
I agree with Will. It'd be one thing if you were calling a method on `paygManager` and it happened to emit a signal (especially after some other work) but it looks like `paygManager` is emitting a signal that should really belong to `PaygUnlockScreen`.